### PR TITLE
net: Consistent checksum handling

### PR DESCRIPTION
**Pull request by laanwj**
In principle, the checksums of P2P packets are simply 4-byte blobs which are the first bytes of SHA256(SHA256(payload)).

Currently they are handled as little-endian 32-bit integers half of the time, as blobs the other half, sometimes copying the one to the other, resulting in somewhat confused code.

This PR changes the handling to be consistent both at packet creation and receiving, making it (I think) easier to understand.

### CODE REVIEW

**Comment by sipa** [comment]
utACK 0a1229617d647d8d288f2afa3006d6a6a5d942a1. This inconsistency has bothered me before.

**Comment by laanwj** [comment]
I've slightly changed the error message to no longer mention nChecksum, as the name of the field changed and is not relevant for troubleshooting anyway:
```
2016-09-28 10:42:19.666628 ProcessMessages(ping, 8 bytes): CHECKSUM ERROR expected 7ef0ca62 was 12234556
```
In some later pull we'd likely want to move all these messages to the net debug category. I'm also going to add a test later that will actually exercise the P2P network error logic.

**Comment by jonasschnelli** [comment]
utACK 41e58faf043864a64a6db08a8df527fa5fd1ec5b
Any reason to keep `CHECKSUM_SIZE = sizeof(int)` (`protocol.h`) instead of just `CHECKSUM_SIZE = 4`?

File: src/main.cpp [hunk 1]
```diff
@@ -6253,11 +6253,12 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman)
         // Checksum
         CDataStream& vRecv = msg.vRecv;
         uint256 hash = Hash(vRecv.begin(), vRecv.begin() + nMessageSize);
-        unsigned int nChecksum = ReadLE32((unsigned char*)&hash);
-        if (nChecksum != hdr.nChecksum)
+        if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0)
```
Comment (jonasschnelli) [inline src/main.cpp]: A size check of `hash.size() <=  CMessageHeader::CHECKSUM_SIZE` is probably a total overkill...

**Comment by laanwj** [comment]
> Any reason to keep `CHECKSUM_SIZE = sizeof(int)` (`protocol.h`) instead of just `CHECKSUM_SIZE = 4`?

Yes, I thought about that too, we **know** that the checksum is always four bytes and it shouldn't change based on what happens to be the architectures int type. WIll do.

File: src/main.cpp (hunk 1 above)
Comment (laanwj) [inline src/main.cpp]: A compile-time assertion could be added... But I don't think there is any risk of someone increasing `CHECKSUM_SIZE` larger than 32 in practice.

**Review by jonasschnelli — APPROVED** [review]
utACK 305087bdf640a5c0b3eeeb90d33ee255d2456eba

File: src/main.cpp [hunk 2]
```diff
@@ -6253,11 +6253,12 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman)
...
         {
-            LogPrintf("%s(%s, %u bytes): CHECKSUM ERROR nChecksum=%08x hdr.nChecksum=%08x\n", __func__,
-               SanitizeString(strCommand), nMessageSize, nChecksum, hdr.nChecksum);
+            LogPrintf("%s(%s, %u bytes): CHECKSUM ERROR expected %s was %s\n", __func__,
+               SanitizeString(strCommand), nMessageSize,
+               HexStr(hash.begin(), hash.begin()+CMessageHeader::CHECKSUM_SIZE),
+               HexStr(hdr.pchChecksum, hdr.pchChecksum+CMessageHeader::CHECKSUM_SIZE));
```
Comment (rebroad) [inline src/main.cpp]: are these the wrong way around? i.e. the expected checksum is supposed to be displayed before the "was"?

**Pull request merged**